segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -0,0 +1,483 @@
+        return String.format("%s:%s:%d:%s:%d",
+                TRUNCATE_RECORD,
+                segmentName,
+                offset,
+                firstChunkName,
+                startOffset);
+    }
+
+    /**
+     * Bootstrap the metadata about critical storage segments by reading and processing the journal.
+     *
+     * @throws Exception Exception in case of any error.
+     */
+    public void bootstrap() throws Exception {
+        try (val txn = metadataStore.beginTransaction()) {
+            // Keep track of offsets at which chunks were added to the system segments.
+            val chunkStartOffsets = new HashMap<String, Long>();
+
+            // Keep track of offests at which system segments were truncated.
+            // We don't need to apply each truncate operation, only need to apply the final truncate offset.
+            val finalTruncateOffsets = new HashMap<String, Long>();
+            val finalFirstChunkStartsAtOffsets = new HashMap<String, Long>();
+
+            // Step 1: Create metadata records for system segments without any chunk information.
+            for (String systemSegment : systemSegments) {
+                SegmentMetadata segmentMetadata = SegmentMetadata.builder()
+                        .name(systemSegment)
+                        .ownerEpoch(epoch)
+                        .maxRollinglength(segmentRollingPolicy.getMaxLength())
+                        .build();
+                segmentMetadata.setActive(true).setOwnershipChanged(true);
+                segmentMetadata.checkInvariants();
+                txn.create(segmentMetadata);
+                txn.markPinned(segmentMetadata);
+            }
+
+            // Step 2: For each epoch, find the corresponding system journal files, process them and apply operations recorded.
+            applySystemLogOperations(txn, chunkStartOffsets, finalTruncateOffsets, finalFirstChunkStartsAtOffsets);
+
+            // Step 3: Adjust the length of the last chunk.
+            adjustLastChunkLengths(txn);
+
+            // Step 4: Apply the truncate offsets.
+            applyFinalTruncateOffsets(txn, finalTruncateOffsets, finalFirstChunkStartsAtOffsets);
+
+            // Step 5: Finally commit all data.
+            txn.commit(true, true);
+        }
+    }
+
+    /**
+     * Process all systemLog entries to recreate the state of metadata storage system segments.
+     */
+    private void applySystemLogOperations(MetadataTransaction txn, HashMap<String, Long> chunkStartOffsets, HashMap<String, Long> finalTruncateOffsets, HashMap<String, Long> finalFirstChunkStartsAtOffsets) throws ChunkStorageException, IOException, StorageMetadataException {
+        for (int epochToRecover = 0; epochToRecover < epoch; epochToRecover++) {
+            // Start scan with file index 0.
+            int fileIndexToRecover = 0;
+            while (chunkStorage.exists(getSystemJournalFileName(containerId, epochToRecover, fileIndexToRecover))) {
```

Comment:
The call to `exists` is unnecessary. Rework your loop to make use of `getInfo`/`openRead` and stop on a not-exist exception.

Reply:
At the current moment I need debug friendly easy code.
Once I am done with all changes, I do plan to implement a few more optimization at the very end as part of the last PR in this group.